dm/syncer/data_validator.go

```diff
@@ -171,10 +173,16 @@ func NewContinuousDataValidator(cfg *config.SubTaskConfig, syncerObj *Syncer, st
 
 	v.workerCnt = cfg.ValidatorCfg.WorkerCount
 	v.processedRowCounts = make([]atomic.Int64, rowChangeTypeCount)
-	v.workers = make([]*validateWorker, v.workerCnt)
+	v.pendingRowCounts = make([]atomic.Int64, rowChangeTypeCount)
+	v.errorRowCounts = make([]atomic.Int64, errorStateTypeCount)
+	// no pre-allocated space for workers
+	// because checkpoint would try visiting all the workers to persist them, though they are nil
+	// set length to zero to prevent nil pointer panic
+	v.workers = make([]*validateWorker, 0)
```

Comment:
Where does the checkpoint access happen? `v.workers` is not protected by a lock, maybe there's a R/W race.

And I guess you want to avoid it iterated by `for`, the zero value of slice (nil) is OK, no need to set it to an empty slice.

Reply:
> Where does the checkpoint access happen? v.workers is not protected by a lock, maybe there's a R/W race.

https://github.com/pingcap/tiflow/blob/4d6f44dfeadc0c95a29a23b271fe0669d2dbc7a1/dm/syncer/validator_checkpoint.go#L213-L306

There is no data race currently because the method called in `validator_checkpoint` is thread-safe and `workers` won't be accessed concurrently by the checkpoint and the validator. 

The `checkpoint.persist` function will be invoked before the validator starts, in order to fetch persisted state from the database. The purpose of setting the empty slice is to prevent the checkpoint visiting nil worker and calling its method (panic). Otherwise, I have to check if the worker is nil every time before I call worker's methods.

Are there any side-effects if I put off allocating the workers array?

Reply:
you can try delete these four lines